tests/conformance/test_conformance.py

```diff
@@ -0,0 +1,449 @@
+# API call (e.g. storage.objects.get) and values are a list of functions which
+# wrap library methods that implement these calls. There may be multiple values
+# because multiple library methods may use the same call (e.g. get could be a
+# read or just a metadata get).
+method_mapping = {
+    "storage.buckets.delete": [delete_bucket],  # S1 start
+    "storage.buckets.get": [get_bucket, reload_bucket],
+    "storage.buckets.getIamPolicy": [get_iam_policy],
+    "storage.buckets.insert": [create_bucket],
+    "storage.buckets.list": [list_buckets],
+    "storage.buckets.lockRententionPolicy": [],  # lock_retention_policy
+    "storage.buckets.testIamPermission": [get_iam_permissions],
+    "storage.notifications.delete": [delete_notification],
+    "storage.notifications.get": [get_notification],
+    "storage.notifications.list": [list_notifications],
+    "storage.objects.get": [get_blob],
+    "storage.objects.list": [list_blobs],  # S1 end
+    "storage.buckets.patch": [patch_bucket],  # S2 start
+    "storage.buckets.setIamPolicy": [],
+    "storage.buckets.update": [update_bucket],
+    "storage.objects.compose": [],  # compose_blob
+    "storage.objects.copy": [copy_blob, rename_blob],
+    "storage.objects.delete": [delete_blob],
+    "storage.objects.insert": [upload_from_string],
+    "storage.objects.patch": [patch_blob],
+    "storage.objects.rewrite": [rewrite_blob],
+    "storage.objects.update": [update_blob],  # S2 end
+    "storage.notifications.insert": [create_notification],  # S4
+}
+
+########################################################################################################################################
+### Helper Methods for Populating Resources ############################################################################################
+########################################################################################################################################
+
+
+def _populate_resource_bucket(client, resources):
+    bucket = client.bucket(uuid.uuid4().hex)
+    client.create_bucket(bucket)
+    resources["bucket"] = bucket
+
+
+def _populate_resource_object(client, resources):
+    bucket_name = resources["bucket"].name
+    bucket = client.get_bucket(bucket_name)
+    blob = bucket.blob(uuid.uuid4().hex)
+    blob.upload_from_string("hello world")
+    blob.reload()
+    resources["object"] = blob
+
+
+def _populate_resource_notification(client, resources):
+    bucket_name = resources["bucket"].name
+    bucket = client.get_bucket(bucket_name)
+    notification = bucket.notification()
+    notification.create()
+    notification.reload()
+    resources["notification"] = notification
+
+
+def _populate_resource_hmackey(client, resources):
+    hmac_key, secret = client.create_hmac_key(
+        service_account_email=_CONF_TEST_SERVICE_ACCOUNT_EMAIL,
+        project_id=_CONF_TEST_PROJECT_ID,
+    )
+    resources["hmac_key"] = hmac_key
+
+
+resource_mapping = {
```

Comment:
Did you try using pytest fixtures for these? Curious about the pros/cons of that.

Reply:
Currently trying to use pytest fixtures and create individual test cases, WIP

Reply:
@tritone I've made a version using pytest fixtures to populate resources. Given that test cases may modify or delete resources, we're unable to fully utilize the teardown/cleanup benefits that come with pytest.fixtures. Another minus is that given our nested structure, I need to populate ALL resources for each test case, despite that the test case may only need a bucket resouce and nothing else. Therefore, it may not make sense to use pytest fixtures in this case.